src/Stratis.Bitcoin/Utilities/Extensions/TaskExtension.cs

```diff
@@ -0,0 +1,49 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Stratis.Bitcoin.Utilities.Extensions
+{
+    public static class TaskExtensions
+    {
+        /// <summary>
+        /// Allows to cancel awaitable operations with a cancellationToken.
+        /// https://devblogs.microsoft.com/pfxteam/how-do-i-cancel-non-cancelable-async-operations/
+        /// </summary>
+        /// <typeparam name="T">Task return type</typeparam>
+        /// <param name="task">The task.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns></returns>
+        /// <exception cref="OperationCanceledException">Task has been cancelled.</exception>
+        public static async Task<T> WithCancellationAsync<T>(this Task<T> task, CancellationToken cancellationToken)
+        {
+            var tcs = new TaskCompletionSource<bool>();
+
+            using (cancellationToken.Register(s => ((TaskCompletionSource<bool>)s).TrySetResult(true), tcs))
```

Comment:
We have to be careful here this is a classic memory leak possibility

Reply:
where is the memory leak?
It's an `using`, it disposes the token registration automatically
afther the if check